File: clients/aonw_flutter/lib/features/map/read_model/map_view.dart
typedef MapHexCoordinate = ({int col, int row});

enum MapGridLayout { oddQFlatTop }

enum MapTerrain {
  ocean,
  coast,
  lake,
  plains,
  grassland,
  desert,
  tundra,
  snow,
  mountain,
  hills,
  wetlands,
  jungle,
  forest,
  river,
}

enum MapResource {
  wheat,
  fish,
  deer,
  sheep,
  rice,
  cow,
  apple,
  banana,
  citrus,
  gold,
  silver,
  gems,
  silk,
  spices,
  cotton,
  grapes,
  ivory,
  pearls,
  coffee,
  cocoa,
  tobacco,
  sugar,
  iron,
  coal,
  oil,
  aluminium,
  uranium,
  horses,
  marble,
}

enum MapObjectiveType { ruins, strategicPass, holySite, legendaryResource }

final class MapTileView {
  MapTileView({
    required this.coordinate,
    required this.displayTerrain,
    required this.yieldTerrain,
    required List<MapTerrain> movementTerrains,
    required List<MapTerrain> terrainTags,
    required List<MapResource> resources,
    required this.height,
  }) : movementTerrains = List.unmodifiable(movementTerrains),
       terrainTags = List.unmodifiable(terrainTags),
       resources = List.unmodifiable(resources);

  final MapHexCoordinate coordinate;
  final MapTerrain displayTerrain;
  final MapTerrain yieldTerrain;
  final List<MapTerrain> movementTerrains;
  final List<MapTerrain> terrainTags;
  final List<MapResource> resources;
  final int height;
}

final class MapObjectiveView {
  const MapObjectiveView({
    required this.id,
    required this.type,
    required this.coordinate,
    required this.requiredHoldTurns,
    required this.victoryPoints,
    required this.goldPerTurn,
  });

  final String id;
  final MapObjectiveType type;
  final MapHexCoordinate coordinate;
  final int requiredHoldTurns;
  final int victoryPoints;
  final int goldPerTurn;
}

final class MapView {
  MapView({
    required this.mapId,
    required this.contentHash,
    required this.gridLayout,
    required this.cols,
    required this.rows,
    required this.defaultZoom,
    required List<MapTileView> tiles,
    required List<MapObjectiveView> objectives,
  }) : tiles = List.unmodifiable(tiles),
       objectives = List.unmodifiable(objectives) {
    _validateHeader();
    _tilesByCoordinate = Map.unmodifiable(_indexTiles());
    _validateObjectives();
  }

  final String mapId;
  final String contentHash;
  final MapGridLayout gridLayout;
  final int cols;
  final int rows;
  final double defaultZoom;
  final List<MapTileView> tiles;
  final List<MapObjectiveView> objectives;
  late final Map<MapHexCoordinate, MapTileView> _tilesByCoordinate;

  MapTileView? tileAt(MapHexCoordinate coordinate) =>
      _tilesByCoordinate[coordinate];

  bool contains(MapHexCoordinate coordinate) =>
      _tilesByCoordinate.containsKey(coordinate);

  bool isWithinBounds(MapHexCoordinate coordinate) =>
      coordinate.col >= 0 &&
      coordinate.col < cols &&
      coordinate.row >= 0 &&
      coordinate.row < rows;

  Map<MapHexCoordinate, MapTileView> _indexTiles() {
    final indexedTiles = <MapHexCoordinate, MapTileView>{};
    for (final tile in tiles) {
      if (!isWithinBounds(tile.coordinate)) {
        throw FormatException(
          'Map tile is outside bounds: ${tile.coordinate}.',
        );
      }
      if (tile.height < 0 || tile.height > 5) {
        throw FormatException(
          'Map tile height is outside the supported range: ${tile.height}.',
        );
      }
      if (indexedTiles.containsKey(tile.coordinate)) {
        throw FormatException('Duplicate map tile: ${tile.coordinate}.');
      }
      indexedTiles[tile.coordinate] = tile;
    }
    if (indexedTiles.length != cols * rows) {
      throw FormatException(
        'Map tile coverage is incomplete: ${indexedTiles.length} of ${cols * rows}.',
      );
    }
    return indexedTiles;
  }

  void _validateObjectives() {
    final objectiveIds = <String>{};
    for (final objective in objectives) {
      if (objective.id.isEmpty || !objectiveIds.add(objective.id)) {
        throw FormatException('Map objective id is empty or duplicated.');
      }
      if (!isWithinBounds(objective.coordinate)) {
        throw FormatException(
          'Map objective is outside bounds: ${objective.coordinate}.',
        );
      }
      if (objective.requiredHoldTurns < 0 ||
          objective.victoryPoints < 0 ||
          objective.goldPerTurn < 0) {
        throw FormatException('Map objective values cannot be negative.');
      }
    }
  }

  void _validateHeader() {
    if (mapId.isEmpty) throw const FormatException('Map id is empty.');
    if (!RegExp(r'^[0-9a-f]{64}$').hasMatch(contentHash)) {
      throw const FormatException('Map content hash is not a SHA-256 digest.');
    }
    if (cols <= 0 || rows <= 0) {
      throw const FormatException('Map dimensions must be positive.');
    }
    if (!defaultZoom.isFinite || defaultZoom <= 0) {
      throw const FormatException(
        'Map default zoom must be finite and positive.',
      );
    }
  }
}

File: engine/crates/aonw_local_runtime/src/client_protocol/encode/map_view.rs
use aonw_content::{
    GridLayout, MapDocument, MapObjective, MapObjectiveType, ResourceType, TerrainType,
    TileDefinition,
};
use aonw_contracts::CoordinateDto;
use aonw_contracts::client::{
    MapGridLayoutDto, MapObjectiveTypeDto, MapObjectiveViewDto, MapResourceDto, MapTerrainDto,
    MapTileViewDto, MapViewDto,
};
use aonw_domain::HexCoord;

pub(crate) fn map(document: &MapDocument) -> Result<MapViewDto, serde_json::Error> {
    let map = document.map();
    Ok(MapViewDto {
        map_id: map.map_id().to_owned(),
        content_hash: map.content_hash()?.to_string(),
        grid_layout: grid_layout(map.grid_layout()),
        cols: map.cols(),
        rows: map.rows(),
        default_zoom: document.default_zoom(),
        tiles: map.tiles().iter().map(tile).collect(),
        objectives: map.objectives().iter().map(objective).collect(),
    })
}

const fn grid_layout(value: GridLayout) -> MapGridLayoutDto {
    match value {
        GridLayout::OddQFlatTop => MapGridLayoutDto::OddQFlatTop,
    }
}

fn tile(value: &TileDefinition) -> MapTileViewDto {
    MapTileViewDto {
        coordinate: coordinate(value.coordinate()),
        display_terrain: terrain(value.display_terrain()),
        yield_terrain: terrain(value.yield_terrain()),
        movement_terrains: value
            .movement_terrains()
            .iter()
            .copied()
            .map(terrain)
            .collect(),
        terrain_tags: value.terrain_tags().iter().copied().map(terrain).collect(),
        resources: value.resources().iter().copied().map(resource).collect(),
        height: value.height(),
    }
}

fn objective(value: &MapObjective) -> MapObjectiveViewDto {
    MapObjectiveViewDto {
        id: value.id().to_owned(),
        objective_type: objective_type(value.objective_type()),
        coordinate: coordinate(value.coordinate()),
        required_hold_turns: value.required_hold_turns(),
        victory_points: value.victory_points(),
        gold_per_turn: value.gold_per_turn(),
    }
}

pub(super) const fn terrain(value: TerrainType) -> MapTerrainDto {
    match value {
        TerrainType::Ocean => MapTerrainDto::Ocean,
        TerrainType::Coast => MapTerrainDto::Coast,
        TerrainType::Lake => MapTerrainDto::Lake,
        TerrainType::Plains => MapTerrainDto::Plains,
        TerrainType::Grassland => MapTerrainDto::Grassland,
        TerrainType::Desert => MapTerrainDto::Desert,
        TerrainType::Tundra => MapTerrainDto::Tundra,
        TerrainType::Snow => MapTerrainDto::Snow,
        TerrainType::Mountain => MapTerrainDto::Mountain,
        TerrainType::Hills => MapTerrainDto::Hills,
        TerrainType::Wetlands => MapTerrainDto::Wetlands,
        TerrainType::Jungle => MapTerrainDto::Jungle,
        TerrainType::Forest => MapTerrainDto::Forest,
        TerrainType::River => MapTerrainDto::River,
    }
}

pub(super) const fn resource(value: ResourceType) -> MapResourceDto {
    match value {
        ResourceType::Wheat => MapResourceDto::Wheat,
        ResourceType::Fish => MapResourceDto::Fish,
        ResourceType::Deer => MapResourceDto::Deer,
        ResourceType::Sheep => MapResourceDto::Sheep,
        ResourceType::Rice => MapResourceDto::Rice,
        ResourceType::Cow => MapResourceDto::Cow,
        ResourceType::Apple => MapResourceDto::Apple,
        ResourceType::Banana => MapResourceDto::Banana,
        ResourceType::Citrus => MapResourceDto::Citrus,
        ResourceType::Gold => MapResourceDto::Gold,
        ResourceType::Silver => MapResourceDto::Silver,
        ResourceType::Gems => MapResourceDto::Gems,
        ResourceType::Silk => MapResourceDto::Silk,
        ResourceType::Spices => MapResourceDto::Spices,
        ResourceType::Cotton => MapResourceDto::Cotton,
        ResourceType::Grapes => MapResourceDto::Grapes,
        ResourceType::Ivory => MapResourceDto::Ivory,
        ResourceType::Pearls => MapResourceDto::Pearls,
        ResourceType::Coffee => MapResourceDto::Coffee,
        ResourceType::Cocoa => MapResourceDto::Cocoa,
        ResourceType::Tobacco => MapResourceDto::Tobacco,
        ResourceType::Sugar => MapResourceDto::Sugar,
        ResourceType::Iron => MapResourceDto::Iron,
        ResourceType::Coal => MapResourceDto::Coal,
        ResourceType::Oil => MapResourceDto::Oil,
        ResourceType::Aluminium => MapResourceDto::Aluminium,
        ResourceType::Uranium => MapResourceDto::Uranium,
        ResourceType::Horses => MapResourceDto::Horses,
        ResourceType::Marble => MapResourceDto::Marble,
    }
}

pub(super) const fn objective_type(value: MapObjectiveType) -> MapObjectiveTypeDto {
    match value {
        MapObjectiveType::Ruins => MapObjectiveTypeDto::Ruins,
        MapObjectiveType::StrategicPass => MapObjectiveTypeDto::StrategicPass,
        MapObjectiveType::HolySite => MapObjectiveTypeDto::HolySite,
        MapObjectiveType::LegendaryResource => MapObjectiveTypeDto::LegendaryResource,
    }
}

pub(super) const fn coordinate(value: HexCoord) -> CoordinateDto {
    CoordinateDto {
        col: value.col(),
        row: value.row(),
    }
}
